### ISSUE TITLE
Deleting build/compose_reports results in a missing error that does not get fixed when rerunning the task

### ISSUE THREAD

**Issue by matejdro** [j-roskopf/ComposeGuard#41]
Performing following steps:

1. Run `<variant>ComposeCompilerCheck` gradle task
2. Delete `build/compose_reports`
3. Run `<variant>ComposeCompilerCheck` gradle task again

Will result in an `Directory '.../build/compose_reports' not exists` error.

It seems Kotlin plugin does not include this directory as part of the UP-TO-DATE checks. I have made a ticket for them: https://youtrack.jetbrains.com/issue/KT-70509

At first glance, a workaround would be to simply add this folder to the output of the Kotlin tasks:

```kotlin
tasks.withType<KotlinCompile>() {
    outputs.dir(project.layout.buildDirectory.dir("compose_reports"))
}
``` 

However, this will cause **all Kotlin tasks in a module to invalidate each other**. That means that, for example, when main kotlin kotlin task generates `compose_reports`, it would invalidate ksp kotlin task and so on. 

Ideally, all different kotlin tasks in a module should have a different output directory and then maybe a merger task that merges them into a single report?

**Comment by j-roskopf** [comment]
@matejdro this is tangentially related to the other issue you opened up. i was (incorrectly) assigning the output directory of the check task as an input of the check task so it'd re-run correctly in this scenario.

what are your thoughts on this:

```
        project.tasks.withType(KotlinCompile::class.java).configureEach(
            object : Action<KotlinCompile<*>> {
                override fun execute(t: KotlinCompile<*>) {
                    t.outputs.upToDateWhen {
                        // where checkExtension.outputDirectory is the build/compose_reports folder by default
                        checkExtension.outputDirectory.get().exists()
                    }
                }
            }
        )
```

I added ksp to my sandbox app and as a test followed this and it seems like everything was working as expected:

1. run `./gradlew :feature:debugComposeCompilerCheck`
2. run `./gradlew :feature:kspDebugKotlin`
3. delete `build/compose_reports`
4. re-run `./gradlew :feature:debugComposeCompilerCheck`
6. See `build/compose_reports` correctly generate as compile task executed
7. re-run `./gradlew :feature:kspDebugKotlin`
8. see no tasks execute, configuration cache used

**Comment by matejdro** [comment]
This can still be vulnerable to this:

1. Run `./gradlew :feature:debugComposeCompilerCheck`
2. Run `./gradlew :feature:releaseComposeCompilerCheck`
3. Delete `build/compose_reports`
4. Run `./gradlew :feature:debugComposeCompilerCheck`
5. Run `./gradlew :feature:releaseComposeCompilerCheck`

Step 5 will fail because folder will exist, but relevant files in that folder will not.

**Comment by matejdro** [comment]
Also found a similar problem with the build cache, if you want, I can create a separate ticket for this:

1. Run `./gradlew clean`
2. Run `./gradlew :feature:debugComposeCompilerCheck --build-cache`
3. Run `./gradlew clean`
4. Run `./gradlew :feature:debugComposeCompilerCheck --build-cache`

Check will fail with `compose_reports not exists` error

**Comment by j-roskopf** [comment]
> Step 5 will fail because folder will exist, but relevant files in that folder will not.

can rectify that with a bit smarter of a upToDateWhen check

```
                    t.outputs.upToDateWhen {
                        checkExtension.outputDirectory.get().listFiles()?.any {
                            it.name.contains(variant.plus("-composables.txt"))
                        } ?: false
                    }
```

**Comment by j-roskopf** [comment]
> Also found a similar problem with the build cache, if you want, I can create a separate ticket for this:
> 
> 1. Run `./gradlew clean`
> 2. Run `./gradlew :feature:debugComposeCompilerCheck --build-cache`
> 3. Run `./gradlew clean`
> 4. Run `./gradlew :feature:debugComposeCompilerCheck --build-cache`
> 
> Check will fail with `compose_reports not exists` error


created a new issue for this!

**Comment by j-roskopf** [comment]
@matejdro I uploaded a 2.2.3-alpha that I think addresses all open issues. would you be willing to give it a spin and let me know if you run into anything? It's appreciated!

**Comment by matejdro** [comment]
Thanks for looking into it!

It seems I can still do this:

1. Add some unstable composables, so the compiler check should fail
2. Run `./gradlew :feature:debugComposeCompilerCheck` <- this should fail with reporting unstable composables
3. Delete all files in the `build/compose_reports`
4.  Run `./gradlew :feature:debugComposeCompilerCheck` <- this time, task passes even though there are unstable composables inside 

I did some experiments with setting the same `upToDateWhen` for the kotlin compile task and it seems to fix above issue.

**Comment by j-roskopf** [comment]
verified the above is addressed in upcoming 2.2.4-alpha

**Comment by matejdro** [comment]
I can still reproduce above steps wih `2.2.4-alpha`. The issue is that, even after deleting `compose_reports`, `debugComposeCompilerCheck` is still `UP-TO-DATE`.

I think this task should never be up to date when there are reported unstable composables? Even if you run it twice without changes, it should fail both times.

**Comment by j-roskopf** [comment]
I am unable to repro unfortunately using the latest published version 2.2.4-alpha. I have recorded a video, am I performing a different set of steps than you?

https://github.com/user-attachments/assets/d877366c-919c-420b-85f6-70a091322bab

**Comment by j-roskopf** [comment]
I also verified the same steps by running `:feature:debugComposeCompilerCheck` instead of just `debugComposeCompilerCheck`. I also ran the same steps across a few android + non-android modules in my multiplatform sandbox app as well (video above is from a regular android project) and saw the expected behavior as well, so I'm a little stumped.

**Comment by matejdro** [comment]
Yes, your steps are different. My step mentions to delete all files from the folder instead of deleting whole folder.

**Comment by matejdro** [comment]
Also, similarly:

1. Add some unstable composables, so the compiler check should fail
2. Run `./gradlew :feature:debugComposeCompilerGenerate`
3. Run `./gradlew :feature:debugComposeCompilerCheck` <- this should not fail, because unstable composable is now in the baseline
4. Manually edit `my-module/compose_reports/...-composables.txt` and remove unstable composable from it
5.  Run `./gradlew :feature:debugComposeCompilerCheck` <- this time, task passes even though unstable composable is no longer in the baseline

**Comment by j-roskopf** [comment]
> Yes, your steps are different. My step mentions to delete all files from the folder instead of deleting whole folder.

ah okay, I see it now and can repro. Will fix in 2.2.7-alpha

**Comment by j-roskopf** [comment]
> Also, similarly:
> 
> 1. Add some unstable composables, so the compiler check should fail
> 2. Run `./gradlew :feature:debugComposeCompilerGenerate`
> 3. Run `./gradlew :feature:debugComposeCompilerCheck` <- this should not fail, because unstable composable is now in the baseline
> 4. Manually edit `my-module/compose_reports/...-composables.txt` and remove unstable composable from it
> 5. Run `./gradlew :feature:debugComposeCompilerCheck` <- this time, task passes even though unstable composable is no longer in the baseline

fixed in 2.2.7-alpha

**Comment by matejdro** [comment]
Seems to be indeed fixed. Thanks!
